src/blockdata/locktime.rs

```diff
@@ -0,0 +1,735 @@
+// Rust Bitcoin Library
+// Written in 2022 by
+//     Tobin C. Harding <me@tobin.cc>
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+//! Provides type [`LockTime`] that implements the logic around nLockTime/OP_CHECKLOCKTIMEVERIFY.
+//!
+//! There are two types of lock time: lock-by-blockheight and lock-by-blocktime, distinguished by
+//! whether `LockTime < LOCKTIME_THRESHOLD`.
+//!
+
+use core::{mem, fmt};
+use core::cmp::{PartialOrd, Ordering};
+use core::convert::TryFrom;
+use core::str::FromStr;
+use core::num::ParseIntError;
+
+use crate::consensus::encode::{self, Decodable, Encodable};
+use crate::io::{self, Read, Write};
+use crate::prelude::*;
+
+/// The Threshold for deciding whether a lock time value is a height or a time (see [Bitcoin Core]).
+///
+/// `LockTime` values _below_ the threshold are interpreted as block heights, values _above_ (or
+/// equal to) the threshold are interpreted as block times (UNIX timestamp, seconds since epoch).
+///
+/// Bitcoin is able to safely use this value because a block height greater than 500,000,000 would
+/// never occur because it would represent a height in approximately 9500 years. Conversely, block
+/// times under 500,000,000 will never happen because they would represent times before 1986 which
+/// are, for obvious reasons, not useful within the Bitcoin network.
+///
+/// [Bitcoin Core]: https://github.com/bitcoin/bitcoin/blob/9ccaee1d5e2e4b79b0a7c29aadb41b97e4741332/src/script/script.h#L39
+pub const LOCK_TIME_THRESHOLD: u32 = 500_000_000;
+
+/// Packed lock time wraps a [`LockTime`] consensus value i.e., the raw `u32` used by the network.
+///
+/// This type implements an naive ordering based on the `u32`, this is _not_ a semantically correct
+/// ordering for a lock time, hence [`LockTime`] does not implement `Ord`. This type is useful if
+/// you want to use a lock time as part of a struct and wish to derive `Ord`. For all other uses,
+/// consider using [`LockTime`] directly.
```

Comment:
Perhaps worth mentioning performance gain for high-perf applications.

Reply:
Can you write something here please, I can not speak definitely on the topic. I read what you posted above 

> Decreased CPU and memory overhead for applications not processing lock time (just passing it around; e.g. indexers)
Faster deserialization

I thought an enum compiled down to a single word in memory?

Reply:
Suggested documentation:

> This struct may be preferred in performance-critical applications because it's slightly smaller than `LockTime` and has a bit more performant (de)serialization. In particular, this may be relevant when the value is not processed, just passed around. Note however that the difference is super-small, so unless you do something extreme you shouldn't worry about it.

## Explanation

Enums compile to a single word only in specific situations:

* C-like enums
* one-variant enums containing one-word value
* multi-variant enums which have only one variant inhabited and that one contains one-word value
* various cases of layout optimizations mostly involving one variant contains a value that forbids some bit patterns (references containing NULL, `bool`, other c-like enums, `File` on Unix...) and not-too-many C-like variants

In other cases the compiler produces something like this

```rust
// enum Foo { A(u32), B(u32) }
struct Foo {
    discriminant: FooDiscriminant,
    data: FooData,
}

enum FooDiscriminant {
    A,
    B,
}

union FooData {
    a: u32,
    b: u32,
}
```

As you can see, this has one word more than the max size of variant data. Each access (in match) checks `discriminant` to see which value is there and the constructor has to `mov` the value in. In our case we check against `LOCK_TIME_THRESHOLD` in decoding just to create the enum even though in principle we shouldn't need that. Unfortunately we can't explain this to the stable compiler today. Thankfully the cost is miniscule and worth better API for most applications.

In the future if `rustc_layout_scalar_valid_range_start` and `rustc_layout_scalar_valid_range_end` attributes get stabilized we could internally use them (may require `unsafe`) and then the compiler would know that the extra discriminant (and the check) can be removed. I have one trick to cause this today on stable with no `unsafe` but it's unsuitable for large types - at best `u16` might be usable.